processing/src/main/java/org/apache/druid/segment/join/JoinableClause.java

```diff
@@ -0,0 +1,134 @@
+ */
+
+package org.apache.druid.segment.join;
+
+import com.google.common.base.Preconditions;
+import org.apache.druid.java.util.common.IAE;
+
+import javax.annotation.Nullable;
+import java.util.Objects;
+
+/**
+ * Represents everything about a join clause except for the left-hand datasource. In other words, if the full join
+ * clause is "t1 JOIN t2 ON t1.x = t2.x" then this class represents "JOIN t2 ON x = t2.x" -- it does not include
+ * references to the left-hand "t1".
+ */
+public class JoinableClause
+{
+  private final String prefix;
+  private final Joinable joinable;
+  private final JoinType joinType;
+  private final JoinConditionAnalysis condition;
+
+  public JoinableClause(@Nullable String prefix, Joinable joinable, JoinType joinType, JoinConditionAnalysis condition)
+  {
+    this.prefix = prefix != null ? prefix : "";
+    this.joinable = Preconditions.checkNotNull(joinable, "joinable");
+    this.joinType = Preconditions.checkNotNull(joinType, "joinType");
+    this.condition = Preconditions.checkNotNull(condition, "condition");
+  }
+
+  /**
+   * The prefix to apply to all columns from the Joinable.
+   */
+  public String getPrefix()
+  {
+    return prefix;
+  }
+
+  /**
+   * The right-hand Joinable.
+   */
+  public Joinable getJoinable()
+  {
+    return joinable;
+  }
+
+  /**
+   * The type of join: LEFT, RIGHT, INNER, or FULL.
+   */
+  public JoinType getJoinType()
+  {
+    return joinType;
+  }
+
+  /**
+   * The join condition. When referring to right-hand columns, it should include the prefix.
+   */
+  public JoinConditionAnalysis getCondition()
+  {
+    return condition;
+  }
+
+  /**
+   * Returns whether "columnName" can be retrieved from the {@link Joinable} represented by this clause (i.e., whether
+   * it starts with {@code prefix} and has at least one other character beyond that).
+   */
+  public boolean includesColumn(final String columnName)
+  {
+    return columnName.startsWith(prefix) && columnName.length() > prefix.length();
+  }
+
+  /**
+   * Removes our prefix from "columnName". Must only be called if {@link #includesColumn} would have returned true
+   * on this column name.
+   */
+  public String unprefix(final String columnName)
+  {
+    if (includesColumn(columnName)) {
+      return columnName.substring(prefix.length());
+    } else {
+      throw new IAE("Column[%s] does not start with prefix[%s]", columnName, prefix);
+    }
+  }
+
+  @Override
+  public boolean equals(Object o)
+  {
+    if (this == o) {
+      return true;
+    }
+    if (o == null || getClass() != o.getClass()) {
+      return false;
+    }
+    JoinableClause that = (JoinableClause) o;
+    return Objects.equals(prefix, that.prefix) &&
+           Objects.equals(joinable, that.joinable) &&
+           joinType == that.joinType &&
+           Objects.equals(condition, that.condition);
+  }
+
+  @Override
+  public int hashCode()
+  {
+    return Objects.hash(prefix, joinable, joinType, condition);
+  }
```

Comment:
hashCode and equals require tests for maintainability. They are easy to write using utility classes. I think Suneet has added some.

Reply:
I added a test using EqualsVerifier.